indico/web/menu.py

```diff
@@ -84,3 +86,94 @@ def __repr__(self):
     @property
     def active(self):
         return request.endpoint in self.endpoints
+
+
+class SideMenuSection(object):
+    """Defines a side menu section (item set).
+
+    :param title: the title of the section (displayed)
+    :param weight: the "weight" (higher means it shows up first)
+    :param active: whether the section should be shown expanded by default
+    :param visible: whether the section should be shown at all
+    :param event_feature: associated feature (id), section will be hidden
+                          if feature is disabled
+    :param icon: icon that will be displayed next to the section title.
+    """
+
+    def __init__(self, title, weight=-1, active=False, visible=True, event_feature=None, icon=None):
+        self.title = title
+        self._active = active
+        self.visible = visible
+        self.event_feature = event_feature
+        self._items = set()
+        self.icon = 'icon-' + icon
+        self.weight = weight
+
+    def add_item(self, item):
+        self._items.add(item)
+
+    @property
+    def items(self):
+        return sorted(self._items, key=lambda x: (-x.weight, x.title))
+
+    @property
+    def active(self):
+        return self._active or any(item.active for item in self._items)
+
+
+class SideMenuItem(object):
+    """Defines a side menu item.
+
+    :param title: the title of the menu item (displayed)
+    :param url: the URL that the link will point to
+    :param weight: the "weight" (higher means it shows up first)
+    :param active: whether the item will be shown as active by default
+    :param visible: whether the item should be shown at all
+    :param event_feature: associated feature (id), the item will be hidden
+                          if feature is disabled
+    :param section: section the item will be put in
+    :param icon: icon that will be displayed next to the item
+    """
+
+    def __init__(self, title, url, weight=-1, active=False, disabled=False, visible=True, event_feature=None,
+                 section=None, icon=None):
+        self.title = title
+        self.url = url
+        self.active = active
+        self.disabled = disabled
+        self.visible = visible
+        self.event_feature = event_feature
+        self.section = section
+        self.weight = weight
+        self.icon = ('icon-' + icon) if icon else None
+
+
+def menu_from_signals(menu_id, active_item=None, **kwargs):
+    """
+    Build a menu (list of entries) with sections/items.
+
+    Information is provided by specific signals and filtered
+    by menu id.
+    This can be used as a very thin framework for menu
+    handling across the app.
+
+    :param menu_id: menu_id used to filter out signal calls
+    :param active_item: ID of currently active menu item
+    :returns: properly sorted list (taking weights into account)
```

Comment:
Maybe we can add a mass-commit that standardises everything to one of them?
But, OTOH, it's one thing that is easy to miss... if someone one of these days adds a `:returns:` then it will spread like wild fire again. I think this is really a case in which it won't hurt to be tolerant.